c10/core/Scalar.h

```diff
@@ -158,14 +198,68 @@ class C10_API Scalar {
     }
   }
 
+  Scalar(Scalar&& rhs) noexcept : tag(rhs.tag) {
+    moveFrom(std::move(rhs));
+  }
+
+  Scalar(const Scalar& rhs) : Scalar(rhs.v, rhs.tag) {}
+
+  Scalar(c10::SymInt si) : tag(Tag::HAS_si) {
+    v.si = si;
+  }
+
+  C10_ALWAYS_INLINE void moveFrom(Scalar&& rhs) noexcept {
+    if (rhs.tag == Tag::HAS_si) {
+      new (&v.si) c10::SymInt(rhs.v.si);
+      rhs.v.si.release_();
+    } else {
+      v.u = rhs.v.u;
+    }
+    tag = rhs.tag;
+    rhs.clearToInt();
+  }
+
+  void clearToInt() noexcept {
+    v.u.i = 0;
+    tag = Tag::HAS_i;
+  }
+
+  // We can't set v in the initializer list using the
+  // syntax v{ .member = ... } because it doesn't work on MSVC
  private:
+  enum class Tag { HAS_d, HAS_i, HAS_z, HAS_b, HAS_si, HAS_sf };
+
+  union Payload {
+    // See [TriviallyCopyablePayload] in IValue.h,
+    union TriviallyCopyablePayload {
+      TriviallyCopyablePayload() : i(0) {}
+      double d;
+      int64_t i;
+      c10::complex<double> z;
+    } u;
+    Payload() : u() {}
+    ~Payload() {}
+    c10::SymInt si;
```

Comment:
it looks like we will always be storing a `SymIntNode` in a SymInt since if it's a concrete int, we would use `.u.i`. 
This means we will be incurring `is_symbolic` check every time we are destroying a `SymInt` again not sure if it matters for Scalars. 

@ezyang @swolchok